File: packages/php/src/Services/B2CService.php
<?php

namespace Yourdudeken\Mpesa\Services;

use Yourdudeken\Mpesa\Http\Client;
use Yourdudeken\Mpesa\Helpers\Signature;

class B2CService
{
    private $client;
    private $signature;

    public function __construct(Client $client, Signature $signature)
    {
        $this->client = $client;
        $this->signature = $signature;
    }

    public function send($phonenumber, $command_id, $amount, $remarks, $result_url = null, $timeout_url = null, $shortCodeType = 'B2C')
    {
        $url = $this->client->getBaseUrl() . '/mpesa/b2c/v1/paymentrequest';

        $body = [
            'InitiatorName' => config('mpesa.initiator_name'),
            'SecurityCredential' => $this->signature->generateSecurityCredential(),
            'CommandID' => $command_id,
            'Amount' => $amount,
            'PartyA' => config('mpesa.b2c_shortcode'),
            'PartyB' => $this->signature->phoneValidator($phonenumber),
            'Remarks' => $remarks,
            'Occassion' => '',
            'ResultURL' => $this->resolveCallbackUrl($result_url, 'b2c_result_url'),
            'QueueTimeOutURL' => $this->resolveCallbackUrl($timeout_url, 'b2c_timeout_url'),
        ];

        return $this->client->post($url, $body, $shortCodeType);
    }

    public function validated($phonenumber, $command_id, $amount, $remarks, $id_number, $result_url = null, $timeout_url = null, $shortCodeType = 'B2C')
    {
        $url = $this->client->getBaseUrl() . '/mpesa/b2cvalidate/v2/paymentrequest';

        $body = [
            'InitiatorName' => config('mpesa.initiator_name'),
            'SecurityCredential' => $this->signature->generateSecurityCredential(),
            'CommandID' => $command_id,
            'Amount' => $amount,
            'PartyA' => config('mpesa.b2c_shortcode'),
            'PartyB' => $this->signature->phoneValidator($phonenumber),
            'Remarks' => $remarks,
            'Occassion' => '',
            'OriginatorConversationID' => $this->signature->getFormattedTimestamp(),
            'IDType' => '01',
            'IDNumber' => $id_number,
            'ResultURL' => $this->resolveCallbackUrl($result_url, 'b2c_result_url'),
            'QueueTimeOutURL' => $this->resolveCallbackUrl($timeout_url, 'b2c_timeout_url'),
        ];

        return $this->client->post($url, $body, $shortCodeType);
    }

    private function resolveCallbackUrl($paramUrl, $configKey)
    {
        $configUrl = config("mpesa.callbacks.{$configKey}");
        
        if ($paramUrl !== null) {
            return $paramUrl;
        } elseif ($configUrl !== null) {
            return $configUrl;
        }
        
        throw new \Exception("Ensure you have set the {$configKey} in the config or passed as a parameter");
    }
}

File: packages/go/mpesa/b2c_service.go
package mpesa

type B2CService struct {
	httpClient *HTTPClient
	auth       *Auth
	helpers    *Helpers
	baseURL    string
}

func NewB2CService(httpClient *HTTPClient, auth *Auth, config *Config) *B2CService {
	return &B2CService{
		httpClient: httpClient,
		auth:       auth,
		helpers:    NewHelpers(config),
		baseURL:    httpClient.GetBaseURL(),
	}
}

func (s *B2CService) Send(phonenumber string, commandId string, amount int, remarks string, resultURL string, timeoutURL string, shortCodeType string) (map[string]interface{}, error) {
	url := s.baseURL + "/mpesa/b2c/v1/paymentrequest"
	body := map[string]interface{}{
		"InitiatorName":      s.helpers.GetConfig("initiator_name"),
		"SecurityCredential": "",
		"CommandID":          commandId,
		"Amount":             amount,
		"PartyA":             s.helpers.GetConfig("b2c_shortcode"),
		"PartyB":             s.helpers.PhoneValidator(phonenumber),
		"Remarks":            remarks,
		"Occassion":          "",
		"ResultURL":          resultURL,
		"QueueTimeOutURL":    timeoutURL,
	}

	token, err := s.auth.GetAccessToken(shortCodeType)
	if err != nil {
		return nil, err
	}

	return s.httpClient.Post(url, body, token)
}
